.github/workflows/cleanup-caches.yml

```diff
@@ -0,0 +1,43 @@
+##
+# Clean up cache keys that are no longer needed.
+##
+name: Cache Garbage Collection
+
+on:
+  pull_request:
+
+jobs:
+  # Delete cache keys scoped to pull requests.
+  #
+  # When a cache is created by a workflow run triggered on a pull request, the cache is created
+  # for the merge ref (refs/pull/.../merge). Because of this, the cache will have a limited scope
+  # and can only be restored by re-runs of the pull request.
+  #
+  # This performs garbage collection on cache keys associated with a closed PR.
+  pr-garbage-collection:
+    name: Delete PR specific cache keys
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+
+    steps:
+      - name: Find and delete PR specific cache keys
+        uses: actions/github-script@98814c53be79b1d30f795b907e553d8679345975 # v6.4.0
+        with:
+          retries: 2
+          retry-exempt-status-codes: 418
+          script: |
+            const action_caches = await github.rest.actions.getActionsCacheList({
+              owner: context.repo.owner,
+              repo: context.repo.repo
+            });
+
+            // Find the caches related to this pull request.
+            for ( let i = 0; i < action_caches.data.actions_caches.length; i++ ) {
+              if ( '${{ github.ref }}' == action_caches.data.actions_caches[ i ].ref ) {
```

Comment:
I think this is the `$GITHUB_REF` value isn't it? which points to the merge branch of the PR? I think this will change every time update a PR or the target branch (usually `trunk`) updates.

should we be matching based on something like the branch name, or the PR number?

Reply:
According to the [documentation](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context), `github.ref` context will always be `refs/pull/<pr_number>/merge`, which matches the pattern defined in the caching documentation linked in the PR description. I'll drop some information about how I tested this in general directly on the PR.

Reply:
> which matches the pattern defined in the caching documentation

I'm not sure I understand what you are suggesting. I am thinking that a given PR could create "15 caches" and this script is deleting "1 cache" because its only looking at the cache of the merge branch _when the PR merged_, but those merge branches update on two events:

 - the PR's branch is changed in any way
 - the target branch is changed in any way

so every commit, every rebase, every other merged PR, updates these merge branches, which I am guessing means some fraction of them create new cache entires.